src/interrupt/mod.rs

```diff
@@ -0,0 +1,188 @@
+//! in-band messages to replace traditional out-of-band assertion of dedicated interrupt lines.
+//! While more complex to implement in a device, message signaled interrupts have some significant
+//! advantages over pin-based out-of-band interrupt signaling. Message signaled interrupts are
+//! supported in PCI bus since its version 2.2, and in later available PCI Express bus. Some non-PCI
+//! architectures also use message signaled interrupts.
+//!
+//! While IRQ is a term commonly used by Operating Systems when dealing with hardware
+//! interrupts, the IRQ numbers managed by OSes are independent of the ones managed by VMM.
+//! For simplicity sake, the term `Interrupt Source` is used instead of IRQ to represent both pin-based
+//! interrupts and MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+
+pub mod legacy;
+pub mod msi;
+
+pub mod trigger;
+
+use std::fmt::{self, Display};
+use std::io;
+use std::ops::Deref;
+
+use crate::interrupt::trigger::Trigger;
+
+/// Errors associated with handling interrupts
+#[derive(Debug)]
+pub enum Error {
+    /// Operation not supported for this interrupt.
+    OperationNotSupported,
+
+    /// The specified configuration is not valid.
+    InvalidConfiguration,
+
+    /// The interrupt was not enabled.
+    InterruptNotEnabled,
+
+    /// Generic IO error,
+    IOError(io::Error),
+}
+
+/// Reuse std::io::Result to simplify interoperability among crates.
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Interrupt error: ")?;
+        match self {
+            Error::OperationNotSupported => write!(f, "operation not supported"),
+            Error::InvalidConfiguration => write!(f, "invalid configuration"),
+            Error::InterruptNotEnabled => write!(f, "the interrupt was not enabled"),
+            Error::IOError(error) => write!(f, "{}", error),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(e: io::Error) -> Error {
+        Error::IOError(e)
+    }
+}
+
+/// Trait for interrupt producers.
+pub trait Interrupt {
+    /// The type of Trigger object used for notifications by this interrupt.
+    type TriggerType: Trigger;
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self) -> std::result::Result<(), <Self::TriggerType as Trigger>::E>;
+
+    /// Returns an interrupt notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest, by writing to the file returned
```

Comment:
Isn't `notifier()` also used for the vmm code to receive the interrupt notifications? That's how it's implemented in the example.

Reply:
It can work both ways since the notifier contains 2 ends, one for the component that triggers the notification and one for the component being notified. Indeed the example shows only the part that is notified. Here's an example where the notifier is used by the device directly: https://github.com/alsrdn/firecracker/blob/d59bd41bd8748bd3eb53623cbff8f4fe7f627198/src/devices/src/virtio/net/device.rs#L768

Reply:
Thanks for the example. What I meant was that the receiver part is quite important and should be mentioned in the doc comment. 

At least this is how I understand the high-level design. An `Interrupt` object is shared between device and vmm (via the `InterruptSourceGroup`, but the group is merely a collection). The object has two main methods:
- `trigger()` is used by the device to trigger the interrupt and notify vmm
- `notifier()` is used by the vmm to receive the triggered interrupt

Reply:
Yes, an `Interrupt` object is shared by between the device and the vmm. It performs 2 main goals:
- interrupt control plane (through the `enable`, `disable` methods that extend to `update`, `mask`, `unmask` for `MaskableInterrupt` and `ConfigurableInterrupt`)
- interrupt data plane through which an event/notification can be delivered between the device and the vmm (`trigger` and `acknowledge`).

Apart from the vmm and the device crate, we have the Hypervisor which is the one responsible for actually delivering the notification to the guest. The Hypervisor typically has a well defined interface for sending notifications (KVM uses irqfds, Xen uses evtchn etc). an `Interrupt` wraps those mechanisms in a common interface and allows them to be configured. When a device uses `trigger`, the `Interrupt` will use this underlying mechanism to signal the Hypervisor.

This interface alone works well but, there are some use cases where you need access to the underlying mechanism in the device. This is true for any device crate that acts more as a control plane and does the emulation somewhere else. A notable example is VFIO that allows us to register the irqfd so that interrupts follow a fast path that doesn't require going through the VMM.
This is where the `notifier` is needed because it provides a way to access the underlying mechanism for delivering the event to the Hypervisor.

While writing this explanation it occurred to me that the `Trigger` trait only complicates the interface without bringing significant value over what already exists in `Interrupt`. My initial reasoning was that it would make for an easier transition between the crates that already use `Trigger`.
So, I would actually drop the `Trigger` completely and just have an associated `NotifierType` value which is unbounded. @gsserge what do you think?

Reply:
I've removed the `Trigger` completely and added some more details in the documentation.